src/pcp/types.rs

```diff
@@ -103,41 +103,58 @@ impl<F: FieldElement> TryFrom<((), &[F])> for Count<F> {
     }
 }
 
-/// This type represents vectors for which `poly(x) == 0` holds for some polynomial `poly` and each
-/// vector element `x`.
+/// This sum type. Each measurement is a integer in `[0, 2^bits)` and the aggregate is the sum of the measurements.
 ///
-/// This type is "generic" in that it can be used to construct high level types. For example, a
-/// boolean vector can be represented as follows:
+/// The validity circuit is based on the SIMD circuit construction of [[BBCG+19], Theorem 5.3].
 ///
-/// ```
-/// use prio::field::{Field64, FieldElement};
-/// use prio::pcp::types::PolyCheckedVector;
-///
-/// let data = vec![Field64::zero(), Field64::one(), Field64::zero()];
-/// let x = PolyCheckedVector::new_range_checked(data, 0, 2);
-/// ```
-/// The proof technique is based on the SIMD circuit construction of
-/// \[[BBG+19](https://eprint.iacr.org/2019/188), Theorem 5.3\].
+/// [BBCG+19]: https://ia.cr/2019/188
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct PolyCheckedVector<F: FieldElement> {
+pub struct Sum<F: FieldElement> {
     data: Vec<F>,
-    poly: Vec<F>,
+    range_checker: Vec<F>,
 }
 
-impl<F: FieldElement> PolyCheckedVector<F> {
-    /// Returns a `poly`-checked vector where `poly(x) == 0` if and only if `x` is in range
-    /// `[start, end)`. The degree of `poly` is equal to `end`.
-    pub fn new_range_checked(data: Vec<F>, start: usize, end: usize) -> Self {
-        Self {
-            data,
-            poly: poly_range_check(start, end),
+impl<F: FieldElement> Sum<F> {
+    /// Constructs a new summand. The value of `summand` must be in `[0, 2^bits)`.
+    pub fn new(summand: u64, bits: u32) -> Result<Self, PcpError> {
+        let summand = usize::try_from(summand).unwrap();
+        let bits = usize::try_from(bits).unwrap();
+
+        if bits > (size_of::<F::Integer>() << 3) {
+            return Err(PcpError::Value(
+                "bits exceeds bit length of the field's integer representation".to_string(),
+            ));
+        }
+
+        let int = F::Integer::try_from(summand).map_err(|err| {
+            PcpError::Value(format!("failed to convert summand to field: {:?}", err))
+        })?;
+
+        let max = F::Integer::try_from(1 << bits).unwrap();
+        if int >= max {
+            return Err(PcpError::Value(
+                "value of summand exceeds bit length".to_string(),
+            ));
+        }
+
+        let one = F::Integer::try_from(1).unwrap();
+        let mut data: Vec<F> = Vec::with_capacity(bits);
+        for l in 0..bits {
+            let l = F::Integer::try_from(l).unwrap();
+            let w = F::from((int >> l) & one);
+            data.push(w);
```

Comment:
IIUC, we'll end up encoding a single `Field126` value as 128 `Field126` values, each of which is 0 or 1. Why not represent it as a single `Field126`, and use `poly_range_check(0, Field126::MAX)` as `range_checker`? Or use some caller provided upper bound (perhaps they know something about the data they're recording that lets them provide a tighter upper bound).

Reply:
OK, that all makes sense. I have no other comments on this PR, but we should resolve #109 so I'll refrain from approving this for now.